authlib/jose/rfc7516/jwe.py

```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
             payload = decode(payload)
         return {'header': protected, 'payload': payload}
 
+    def deserialize_json(self, obj, key, decode=None, sender_key=None):
+        obj = ensure_dict(obj, 'JWE')
+
+        if 'protected' in obj:
+            protected = extract_header(to_bytes(obj['protected']), DecodeError)
+        else:
+            protected = None
+
+        unprotected = obj.get('unprotected')
+
+        recipients = obj['recipients']
+        for recipient in recipients:
+            recipient['encrypted_key'] = extract_segment(
+                to_bytes(recipient['encrypted_key']), DecodeError, 'encrypted key')
+
+        if 'aad' in obj:
+            jwe_aad = extract_segment(to_bytes(obj['aad']), DecodeError, 'JWE AAD')
+        else:
+            jwe_aad = None
+
+        iv = extract_segment(to_bytes(obj['iv']), DecodeError, 'initialization vector')
+
+        ciphertext = extract_segment(to_bytes(obj['ciphertext']), DecodeError, 'ciphertext')
+
+        tag = extract_segment(to_bytes(obj['tag']), DecodeError, 'authentication tag')
+
+        shared_header = JWESharedHeader(protected, unprotected)
+
+        alg = self.get_header_alg(shared_header)
+        enc = self.get_header_enc(shared_header)
+        zip_alg = self.get_header_zip(shared_header)
+
+        self._validate_sender_key(sender_key, alg)
+        self._validate_private_headers(shared_header, alg)
+        for recipient in recipients:
+            self._validate_private_headers(recipient['header'], alg)
+
+        key = alg.prepare_key(key)
+        if sender_key is not None:
+            sender_key = alg.prepare_key(sender_key)
+
+        def _unwrap_with_sender_key_and_tag(ek, header):
+            return alg.unwrap(enc, ek, header, key, sender_key, tag)
+
+        def _unwrap_with_sender_key_and_without_tag(ek, header):
+            return alg.unwrap(enc, ek, header, key, sender_key)
+
+        def _unwrap_without_sender_key_and_tag(ek, header):
+            return alg.unwrap(enc, ek, header, key)
+
+        def _unwrap_for_some_recipient(unwrap_func):
+            for recipient in recipients:
```

Comment:
I believe we must be able to provide a key with a key ID, so that we can avoid that iteration, and try to decrypt for the given key only.

Reply:
Added such logic.